src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -180,27 +180,47 @@ void ExtentPlacementManager::init(
   writer_refs.clear();
 
   ceph_assert(RECLAIM_GENERATIONS > 0);
-  auto segment_cleaner = dynamic_cast<SegmentCleaner*>(cleaner.get());
-  ceph_assert(segment_cleaner != nullptr);
   data_writers_by_gen.resize(RECLAIM_GENERATIONS, {});
-  for (reclaim_gen_t gen = 0; gen < RECLAIM_GENERATIONS; ++gen) {
-    writer_refs.emplace_back(std::make_unique<SegmentedOolWriter>(
-          data_category_t::DATA, gen, *segment_cleaner,
-          segment_cleaner->get_ool_segment_seq_allocator()));
-    data_writers_by_gen[gen] = writer_refs.back().get();
-  }
-
   md_writers_by_gen.resize(RECLAIM_GENERATIONS - 1, {});
-  for (reclaim_gen_t gen = 1; gen < RECLAIM_GENERATIONS; ++gen) {
-    writer_refs.emplace_back(std::make_unique<SegmentedOolWriter>(
-          data_category_t::METADATA, gen, *segment_cleaner,
-          segment_cleaner->get_ool_segment_seq_allocator()));
-    md_writers_by_gen[gen - 1] = writer_refs.back().get();
-  }
+  if (trimmer->get_journal_type() == journal_type_t::SEGMENTED) {
+    auto segment_cleaner = dynamic_cast<SegmentCleaner*>(cleaner.get());
+    ceph_assert(segment_cleaner != nullptr);
+    for (reclaim_gen_t gen = 0; gen < RECLAIM_GENERATIONS; ++gen) {
+      writer_refs.emplace_back(std::make_unique<SegmentedOolWriter>(
+	    data_category_t::DATA, gen, *segment_cleaner,
+	    segment_cleaner->get_ool_segment_seq_allocator()));
+      data_writers_by_gen[gen] = writer_refs.back().get();
+    }
 
-  for (auto *device : segment_cleaner->get_segment_manager_group()
-                                     ->get_segment_managers()) {
-    add_device(device);
+    for (reclaim_gen_t gen = 1; gen < RECLAIM_GENERATIONS; ++gen) {
+      writer_refs.emplace_back(std::make_unique<SegmentedOolWriter>(
+	    data_category_t::METADATA, gen, *segment_cleaner,
+	    segment_cleaner->get_ool_segment_seq_allocator()));
+      md_writers_by_gen[gen - 1] = writer_refs.back().get();
+    }
+
+    for (auto *device : segment_cleaner->get_segment_manager_group()
+				       ->get_segment_managers()) {
+      add_device(device);
+    }
+  } else {
+    assert(trimmer->get_journal_type() == journal_type_t::RANDOM_BLOCK);
+    auto rb_cleaner = dynamic_cast<RBMCleaner*>(cleaner.get());
+    ceph_assert(rb_cleaner != nullptr);
+    // all generations have the same writer for now
+    writer_refs.emplace_back(std::make_unique<RandomBlockOolWriter>(
+	    rb_cleaner));
+    for (reclaim_gen_t gen = 0; gen < RECLAIM_GENERATIONS; ++gen) {
```

Comment:
Let's only register the `OOL_GENERATION` data/md_writers for `RBCleaner`, and assert that all the other writers are impossible to be accessed in the write path if with `RBCleaner`.